### PR TITLE
Make every GPU stick to its shard

### PR DESCRIPTION
**Pull request by JanuszL**
- makes every shard_id to have the same shard of data set across all epochs

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 1]
```diff
@@ -38,7 +38,9 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
     }
 
   void ReadSample(Tensor<CPUBackend>* tensor) override {
-    if (current_index_ == indices_.size()) {
+    if ((current_index_ == indices_.size()) ||
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: extract this so that it can be reused in both loaders?

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 2]
```diff
@@ -103,24 +105,22 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
     DALI_ENFORCE(!uris_.empty(), "No files specified.");
     std::vector<std::string> index_uris = options.GetRepeatedArgument<std::string>("index_path");
     ReadIndexFile(index_uris);
-    size_t num_indices = indices_.size();
-    current_index_ = start_index(shard_id_, num_shards_, num_indices);
-    int64 seek_pos, size;
-    std::tie(seek_pos, size, current_file_index_) = indices_[current_index_];
-    current_file_ = FileStream::Open(uris_[current_file_index_], read_ahead_);
-    current_file_->Seek(seek_pos);
+    current_file_index_ = -1;
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: constant saying what -1 means

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 3]
```diff
@@ -103,24 +105,22 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
...
-  void Reset() {
-    current_index_ = 0;
+  void Reset() override {
     int64 seek_pos, size;
     size_t file_index;
+    current_index_ = start_index(shard_id_, num_shards_, indices_.size());
     std::tie(seek_pos, size, file_index) = indices_[current_index_];
     if (file_index != current_file_index_) {
-      current_file_->Close();
+      if (current_file_index_ != static_cast<size_t>(-1)) {
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: use constant here

File: dali/pipeline/operators/reader/loader/sequence_loader.h [hunk 4]
```diff
@@ -93,27 +93,28 @@ class SequenceLoader : public Loader<CPUBackend, TensorSequence> {
...
     mmap_reserver = FileStream::FileStreamMappinReserver(
         static_cast<unsigned int>(initial_buffer_fill_) * sequence_length_);
     copy_read_data_ = !mmap_reserver.CanShareMappedData();
-    if (shuffle_) {
-      // seeded with hardcoded value to get
-      // the same sequence on every shard
-      std::mt19937 g(524287);
-      std::shuffle(sequences_.begin(), sequences_.end(), g);
-    }
+    // seeded with hardcoded value to get
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/sequence_loader.h]: you dropped the `if (shuffle_)`. Was it intentional?

File: dali/test/python/test_data_containers.py [hunk 5]
```diff
@@ -87,7 +89,7 @@ def define_graph(self):
 class COCOReaderPipeline(CommonPipeline):
     def __init__(self, batch_size, num_threads, device_id, num_gpus, data_paths):
         super(COCOReaderPipeline, self).__init__(batch_size, num_threads, device_id)
-        self.input = ops.COCOReader(file_root = data_paths[0], annotations_file=data_paths[1])
+        self.input = ops.COCOReader(file_root = data_paths[0], annotations_file=data_paths[1], shard_id = device_id, num_shards = num_gpus)
```
Comment (jantonguirao) [inline dali/test/python/test_data_containers.py]: maybe break lines

File: dali/pipeline/operators/reader/loader/sequence_loader.h (hunk 4 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/sequence_loader.h]: Yes. It was wrong from the very beginning. We should always do initial shuffling to mix data, so GPU0 won't get only 0 class data for a few initial batches.

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Done

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 2 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Done

File: dali/test/python/test_data_containers.py (hunk 5 above)
Comment (JanuszL) [reply dali/test/python/test_data_containers.py]: Done

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 3 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Done

**Comment by JanuszL** [comment]
Builds 634069

File: dali/pipeline/operators/reader/loader/sequence_loader.h (hunk 4 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/sequence_loader.h]: Grr, it breaks test. Video loader is no better. Reverting to original code. It need to be fixed separately.

File: dali/pipeline/operators/reader/loader/loader.h [hunk 6]
```diff
@@ -167,6 +167,15 @@ class Loader {
   virtual Index Size() = 0;
 
  protected:
+  // Reset reader to the first sample
+  virtual void Reset() = 0;
+
+  // Check if given reader moved to the next shard
+  virtual bool IfNextShard(Index current_index) {
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/loader.h]: Maybe `IsNextShard` would be a better name?

File: dali/pipeline/operators/reader/loader/lmdb.h [hunk 7]
```diff
@@ -122,13 +115,25 @@ class LMDBReader : public Loader<CPUBackend, Tensor<CPUBackend>> {
   }
 
  private:
+  void Reset() override {
+    // work out how many entries to move forward to handle sharding
+    current_index_ = start_index(shard_id_, num_shards_, Size());
+    bool ok = lmdb::SeekLMDB(mdb_cursor_, MDB_FIRST, &key_, &value_);
+    DALI_ENFORCE(ok, "lmdb::SeekLMDB failed");
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/lmdb.h]: Slighlty out of the scope of this PR but I think this error should be more explicit. Something like:
```suggestion
    DALI_ENFORCE(ok, "lmdb::SeekLMDB failed on MDB_FIRST with arguments key:"
                                 + to_string(key_) + " value:"  to_string(value_));
```

File: dali/pipeline/operators/reader/loader/lmdb.h [hunk 8]
```diff
@@ -122,13 +115,25 @@ class LMDBReader : public Loader<CPUBackend, Tensor<CPUBackend>> {
...
+
+    for (size_t i = 0; i < current_index_; ++i) {
+      bool ok = lmdb::SeekLMDB(mdb_cursor_, MDB_NEXT, &key_, &value_);
+      DALI_ENFORCE(ok, "lmdb::SeekLMDB failed");
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/lmdb.h]: Same here

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 9]
```diff
@@ -134,6 +132,7 @@ class IndexedFileLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
   size_t current_file_index_;
   std::unique_ptr<FileStream> current_file_;
   FileStream::FileStreamMappinReserver mmap_reserver;
+  static constexpr int INVALID_INDEX = -1;
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: Why static?

File: dali/pipeline/operators/reader/loader/sequence_loader.h (hunk 4 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/sequence_loader.h]: My bad, it looks we are doing the same for other reader and it was meant to be like that.

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h (hunk 9 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: constexpr members need to be static.

**Comment by JanuszL** [comment]
Rebuilds 634144

**Comment by JanuszL** [comment]
RN50 for MXNet, PyTorch and TensorFlow converges.

File: dali/pipeline/operators/reader/loader/file_loader.cc [hunk 10]
```diff
@@ -98,8 +98,8 @@ void FileLoader::ReadSample(ImageLabelWrapper* image_label) {
   auto image_pair = image_label_pairs_[current_index_++];
 
   // handle wrap-around
-  if (current_index_ == Size()) {
-    current_index_ = 0;
+  if (IsNextShard(current_index_)) {
+    Reset();
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/file_loader.cc]: is reset equivalent to current_index_ = 0 ?

File: dali/pipeline/operators/reader/loader/loader.h [hunk 11]
```diff
@@ -167,6 +167,15 @@ class Loader {
   virtual Index Size() = 0;
 
  protected:
+  // Reset reader to the first sample
+  virtual void Reset() = 0;
+
+  // Check if given reader moved to the next shard
+  virtual bool IsNextShard(Index current_index) {
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/loader.h]: inline?

File: dali/pipeline/operators/reader/loader/file_loader.cc (hunk 10 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/file_loader.cc]: No, the whole point of that patch is to stick every GPU to its shard. So reset will assign current_index_ to appropriate value, usually != 0.

File: dali/pipeline/operators/reader/loader/loader.h (hunk 11 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/loader.h]: Done
